File: src/services/csv_processor.py
import pandas as pd
from typing import Dict, List
from datetime import datetime

class CSVProcessor:
    # Known CSV formats for different banks
    BANK_FORMATS = {
        'chase': ['Date', 'Description', 'Amount'],
        'bank_of_america': ['Posted Date', 'Payee', 'Amount'],
        'wells_fargo': ['Date', 'Merchant', 'Amount'],
    }

    def __init__(self):
        self.supported_banks = list(self.BANK_FORMATS.keys())

    def detect_bank_format(self, df: pd.DataFrame) -> str:
        """Detect which bank format the CSV matches"""
        columns = set(df.columns)
        
        for bank, expected_columns in self.BANK_FORMATS.items():
            if all(col in columns for col in expected_columns):
                return bank
                
        raise ValueError("Unsupported CSV format")

    def normalize_columns(self, df: pd.DataFrame, bank_format: str) -> pd.DataFrame:
        """Standardize column names based on bank format"""
        if bank_format == 'chase':
            return df.rename(columns={
                'Date': 'date',
                'Description': 'merchant',
                'Amount': 'amount'
            })
        elif bank_format == 'bank_of_america':
            return df.rename(columns={
                'Posted Date': 'date',
                'Payee': 'merchant',
                'Amount': 'amount'
            })
        # Add more bank formats as needed
        
        return df

    def process_csv(self, file_path: str) -> List[Dict]:
        """Process CSV file and return normalized transactions"""
        try:
            # Read CSV file
            df = pd.read_csv(file_path)
            
            # Detect bank format
            bank_format = self.detect_bank_format(df)
            
            # Normalize columns
            df = self.normalize_columns(df, bank_format)
            
            # Clean and standardize data
            df['date'] = pd.to_datetime(df['date']).dt.strftime('%Y-%m-%d')
            df['amount'] = df['amount'].astype(float)
            df['merchant'] = df['merchant'].str.strip().str.upper()
            
            # Convert to list of dictionaries
            transactions = df.to_dict('records')
            
            return transactions
            
        except Exception as e:
            raise Exception(f"Error processing CSV: {str(e)}") 

File: src/services/csv_processor.ts
import Papa from 'papaparse';

interface Transaction {
  date: string;
  merchant: string;
  amount: number;
  category?: string;
  bank?: string;
}

export class CSVProcessor {
  private BANK_FORMATS = {
    'american_express': [
      'Date', 'Description', 'Amount', 'Extended Details', 
      'Appears On Your', 'Address', 'City/State', 'Zip Code',
      'Country', 'Reference', 'Category', 'Bank'
    ]
  };

  processCSV(csvContent: string): Transaction[] {
    const results = Papa.parse(csvContent, {
      header: true,
      skipEmptyLines: true
    });

    if (results.errors.length > 0) {
      throw new Error(`CSV parsing error: ${results.errors[0].message}`);
    }

    const data = results.data;
    const headers = Object.keys(data[0]);
    
    // Detect bank format
    const bankFormat = this.detectBankFormat(headers);
    
    return this.transformData(data, bankFormat);
  }

  private detectBankFormat(headers: string[]): string {
    // Simple detection based on headers matching
    if (headers.includes('Extended Details')) {
      return 'american_express';
    }
    
    throw new Error('Unsupported CSV format');
  }

  private transformData(data: any[], bankFormat: string): Transaction[] {
    switch (bankFormat) {
      case 'american_express':
        return data.map(row => ({
          date: row['Date'],
          merchant: row['Description'],
          amount: this.parseAmount(row['Amount']),
          category: row['Category'],
          bank: 'American Express'
        }));
      default:
        throw new Error(`Unsupported bank format: ${bankFormat}`);
    }
  }

  private parseAmount(amount: string): number {
    // Remove currency symbols and convert to number
    const cleanAmount = amount.replace(/[$,]/g, '');
    return parseFloat(cleanAmount);
  }
} 